beacon_chain/state/state_transition.py

```diff
@@ -322,6 +324,64 @@ def fill_recent_block_hashes(active_state: ActiveState,
     )
 
 
+def compute_cycle_transitions(
+        crystallized_state: CrystallizedState,
+        active_state: ActiveState,
+        block: 'Block',
+        config: Dict[str, Any]=DEFAULT_CONFIG) -> Tuple[CrystallizedState, ActiveState]:
+    while block.slot_number >= crystallized_state.last_state_recalc + config['cycle_length']:
+        crystallized_state, active_state = initialize_new_cycle(
+            crystallized_state,
+            active_state,
+            block,
+            config=config,
+        )
+    return crystallized_state, active_state
+
+
+def ready_for_dynasty_change(crystallized_state: CrystallizedState,
+                             block: 'Block',
+                             config: Dict[str, Any]=DEFAULT_CONFIG) -> bool:
+    # gather every shard in shard_and_committee_for_slots
+    required_shards = set()
+    for shard_and_committee_for_slot in crystallized_state.shard_and_committee_for_slots:
+        for shard_and_committee in shard_and_committee_for_slot:
+            required_shards.add(shard_and_committee.shard_id)
+
+    # check that crosslinks were updated for all required shards
+    required_crosslinks_updated = True
+    for shard_id, crosslink in enumerate(crystallized_state.crosslink_records):
+        if shard_id in required_shards:
+            required_crosslinks_updated &= crosslink.slot > crystallized_state.dynasty_start
+
+
+    slots_since_last_dynasty_change = block.slot_number - crystallized_state.dynasty_start
+
+    return slots_since_last_dynasty_change >= config['min_dynasty_length'] and \
+           crystallized_state.last_finalized_slot > crystallized_state.dynasty_start and \
+           required_crosslinks_updated
+
+
+def compute_dynasty_transition(crystallized_state: CrystallizedState,
+                               block: 'Block',
+                               config: Dict[str, Any]=DEFAULT_CONFIG) -> CrystallizedState:
+    crystallized_state = deepcopy(crystallized_state)
+    crystallized_state.current_dynasty += 1
+
+    # Not current in spec, but should be added soon
+    crystallized_state.dynasty_start = crystallized_state.last_state_recalc
```

Comment:
The spec is updated to `Set last_dynasty_start = block.slot_number`
(should be `crystallized_state.dynasty_start`)

Reply:
spec changed back after mini debate with V :)